docs/enhancements/saas-per-organization-feature-access.md

```diff
@@ -0,0 +1,79 @@
+---
+title: saas-per-organization-feature-access
+authors:
+  - "@nmagnezi"
+creation-date: 2022-05-02
+last-updated: 2022-05-02
+---
+
+# SaaS Per-Organization Feature Access
+
+The current OCM-based authorization scheme used by assisted service as a part of the SaaS operation does not support exposing features to certain users.
+We wish to allow the following:
+* An ability to expose selected group(s) of users to assisted-service functionality (in OCM, the term is `capability`).
+* The user grouping should be based on the organization ID.
+
+
+## Motivation
+
+Some features might be of interest to partners or early adopters.
+Other features might not be ready for public consumption: tech-preview or an internal-only functionality.
+Exposing such features to a pre-determined group of users can help us to gather feedback and improve while keeping the other users with a production-level user experience.
+
+
+### Goals
+* Add a generic method to assisted service authorization middleware to assert against an organization's capability. The method should query OCM based on the payload and asserted capability name.
+* Implement a functionality-specific assertion for `ARM`.
+
+
+## Proposal
+
+
+Initially, the authorization middleware should have a method to check if the user belongs to an organization with a particular capability.
+Adding such a method would be the first step, both to assert organization capabilities, such as ARM support, and any other capability of interest in the future.
+
+The API endpoints that need ARM support feature enablement:
+
+### V2ListSupportedOpenshiftVersions API
+
+* [ReleaseImage](../../models/release_image.go) has a `CPUArchitecture` attribute.
+* Expected to filter out ARM version if that capability is not included for the user organization.
+
```

Comment:
Maybe we should add an API like /v2/feature-gates. I.e. returns a list of available features for the current user. That way, both API user and UI could invoke a single query to get all capabilities. We could also cache the list in the service if makes sense.

Reply:
Isn't this duplicating a check of user organization capabilities in AMS/OCM API? both assisted-service and the UI should query that API.
IF we implement `/v2/feature-gate` in assisted-service, won't we just proxy that AMS/OCM API?

Reply:
@avishayt based on the meeting that we had and decided that there are 2 different APIs: one for the supported features per OCP version and the other for the capabilities, do we need to filter the list as Nir wrote here?

Reply:
I believe we said we would have a single API as we have today (feature support levels).  Capabilities should be queried from AMS.
I assumed that users without capabilities should see no trace of a feature that they can't use.  So for example, if a user isn't authorized to use ARM then the behavior should be as if the service doesn't even recognize it as an option (i.e., the same behavior as specifying "cpu_architecture=foobar" - "foobar" doesn't show up in the list of architectures and APIs return 400 if that's specified).  Alternatively, we can decide to expose the existence of the ARM feature to users but not allow setting it, but this seems less user-friendly to me.